File: ui/adaptation.js
(() => {
  const publicPlay = play => play.public || play;
  const beliefs = play => publicPlay(play).beliefs || play.offense_observed?.belief_after || {};
  const graphCardIds = play => publicPlay(play).graph_card_ids || [];
  const offenseCall = play => publicPlay(play).offense_action?.concept_family || null;
  const defenseCall = play => publicPlay(play).defense_action?.coverage_family || null;
  const cardMap = graphCards => Array.isArray(graphCards)
    ? Object.fromEntries(graphCards.map(card => [card.id, card]))
    : graphCards;

  function classifyAdaptationReasons(plays, graphCards) {
    const cards = cardMap(graphCards || {});
    const seen = new Set();
    const reasons = {};

    plays.forEach((play, index) => {
      const currentCards = graphCardIds(play);
      if (currentCards.some(id => !seen.has(id))) {
        reasons[index] = 'graph-fire';
      }

      if (index > 0 && !reasons[index]) {
        const priorBeliefs = beliefs(plays[index - 1]);
        const currentBeliefs = beliefs(play);
        const keys = new Set([...Object.keys(priorBeliefs), ...Object.keys(currentBeliefs)]);
        for (const key of keys) {
          if (Math.abs(Number(currentBeliefs[key] || 0) - Number(priorBeliefs[key] || 0)) >= 0.10) {
            reasons[index] = 'belief-shift';
            break;
          }
        }
      }

      if (index > 0 && !reasons[index]) {
        const counters = new Set(graphCardIds(plays[index - 1]).flatMap(id => cards[id]?.counters || []));
        const offenseCounter = offenseCall(play) !== offenseCall(plays[index - 1]) && counters.has(offenseCall(play));
        const defenseCounter = defenseCall(play) !== defenseCall(plays[index - 1]) && counters.has(defenseCall(play));
        if (offenseCounter || defenseCounter) reasons[index] = 'counter-call';
      }

      currentCards.forEach(id => seen.add(id));
    });

    return reasons;
  }

  function classifyAdaptation(plays, graphCards) {
    return new Set(Object.keys(classifyAdaptationReasons(plays, graphCards)).map(Number));
  }

  window.CBAdaptation = { classifyAdaptation, classifyAdaptationReasons };
})();

File: engine/coachbench/adaptation.py
from __future__ import annotations

from typing import Any


REASON_ORDER = ("graph-fire", "belief-shift", "counter-call")


def _public(play: dict[str, Any]) -> dict[str, Any]:
    return play.get("public", play)


def _beliefs(play: dict[str, Any]) -> dict[str, float]:
    public = _public(play)
    beliefs = public.get("beliefs")
    if isinstance(beliefs, dict):
        return beliefs
    observed = play.get("offense_observed", {})
    beliefs = observed.get("belief_after")
    return beliefs if isinstance(beliefs, dict) else {}


def _offense_call(play: dict[str, Any]) -> str | None:
    action = _public(play).get("offense_action", {})
    return action.get("concept_family")


def _defense_call(play: dict[str, Any]) -> str | None:
    action = _public(play).get("defense_action", {})
    return action.get("coverage_family")


def _cards(play: dict[str, Any]) -> list[str]:
    return list(_public(play).get("graph_card_ids") or [])


def _card_map(graph_cards: dict[str, Any] | list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    if isinstance(graph_cards, list):
        return {card["id"]: card for card in graph_cards}
    return graph_cards


def classify_adaptation_reasons(
    plays: list[dict[str, Any]],
    graph_cards: dict[str, Any] | list[dict[str, Any]],
) -> dict[int, str]:
    cards_by_id = _card_map(graph_cards)
    seen_cards: set[str] = set()
    reasons: dict[int, str] = {}

    for index, play in enumerate(plays):
        current_cards = _cards(play)
        first_fire = any(card_id not in seen_cards for card_id in current_cards)
        if first_fire:
            reasons[index] = "graph-fire"

        if index > 0 and index not in reasons:
            prior_beliefs = _beliefs(plays[index - 1])
            current_beliefs = _beliefs(play)
            for key in set(prior_beliefs) | set(current_beliefs):
                if abs(float(current_beliefs.get(key, 0)) - float(prior_beliefs.get(key, 0))) >= 0.10:
                    reasons[index] = "belief-shift"
                    break

        if index > 0 and index not in reasons:
            prior_cards = _cards(plays[index - 1])
            counters = {
                counter
                for card_id in prior_cards
                for counter in cards_by_id.get(card_id, {}).get("counters", [])
            }
            if counters:
                prior_offense, current_offense = _offense_call(plays[index - 1]), _offense_call(play)
                prior_defense, current_defense = _defense_call(plays[index - 1]), _defense_call(play)
                offense_counter = current_offense != prior_offense and current_offense in counters
                defense_counter = current_defense != prior_defense and current_defense in counters
                if offense_counter or defense_counter:
                    reasons[index] = "counter-call"

        seen_cards.update(current_cards)

    return reasons


def classify_adaptation(
    plays: list[dict[str, Any]],
    graph_cards: dict[str, Any] | list[dict[str, Any]],
) -> set[int]:
    return set(classify_adaptation_reasons(plays, graph_cards))
